modules/utxo-lib/src/testutil/transaction.ts

```diff
@@ -0,0 +1,183 @@
+
+/**
+ * create unspent object from input script type, index, network and root wallet key.
+ */
+export function toTxnUnspent<TNumber extends number | bigint>(
+  input: TxnInput<TNumber>,
+  index: number,
+  network: Network,
+  rootWalletKeys: RootWalletKeys
+): Unspent<TNumber> {
+  if (input.scriptType === 'p2shP2pk') {
+    return mockReplayProtectionUnspent<TNumber>(network, input.value, { key: rootWalletKeys['user'], vout: index });
+  } else {
+    return mockWalletUnspent<TNumber>(network, input.value, {
+      chain: getInternalChainCode(input.scriptType),
+      vout: index,
+      keys: rootWalletKeys,
+      index,
+    });
+  }
+}
+
+/**
+ * returns signer and cosigner names for TxnInputScriptType.
+ * user and undefined as signer and cosigner respectively for p2shP2pk.
+ * user and bitgo as signer and cosigner respectively for other input script types.
+ */
+export function getTxnSigners(inputType: TxnInputScriptType): { signerName: KeyName; cosignerName?: KeyName } {
+  return {
+    signerName: 'user',
+    cosignerName: inputType === 'p2shP2pk' ? undefined : 'bitgo',
+  };
+}
+
+/**
+ * signs with first or second signature for single input.
+ * p2shP2pk is signed only with first sign.
+ */
+export function signTxnInput<TNumber extends number | bigint>(
+  txb: UtxoTransactionBuilder<TNumber>,
+  input: TxnInput<TNumber>,
+  inputIndex: number,
+  rootWalletKeys: RootWalletKeys,
+  sign: 'halfsigned' | 'fullsigned'
+): void {
+  const { signerName, cosignerName } = getTxnSigners(input.scriptType);
+  const unspent = toTxnUnspent(input, inputIndex, txb.network, rootWalletKeys);
+  if (sign === 'halfsigned') {
+    if (input.scriptType === 'p2shP2pk') {
+      signInputP2shP2pk(txb, inputIndex, rootWalletKeys[signerName]);
+    } else if (isWalletUnspent(unspent) && cosignerName) {
+      signInputWithUnspent(
+        txb,
+        inputIndex,
+        unspent,
+        WalletUnspentSigner.from(rootWalletKeys, rootWalletKeys[signerName], rootWalletKeys[cosignerName])
+      );
+    }
+  }
+  if (isWalletUnspent(unspent) && sign === 'fullsigned' && cosignerName) {
+    signInputWithUnspent(
+      txb,
+      inputIndex,
+      unspent,
+      WalletUnspentSigner.from(rootWalletKeys, rootWalletKeys[cosignerName], rootWalletKeys[signerName])
+    );
+  }
+}
+
+/**
+ * signs with first or second signature for all inputs.
+ * p2shP2pk is signed only with first sign.
+ */
+export function signAllTxnInputs<TNumber extends number | bigint>(
+  txb: UtxoTransactionBuilder<TNumber>,
+  inputs: TxnInput<TNumber>[],
+  rootWalletKeys: RootWalletKeys,
+  sign: 'halfsigned' | 'fullsigned'
+): void {
+  inputs.forEach((input, index) => {
+    signTxnInput(txb, input, index, rootWalletKeys, sign);
+  });
+}
+
+/**
+ * construct transaction for given inputs, outputs, network and root wallet keys.
+ */
+export function constructTxnBuilder<TNumber extends number | bigint>(
```

Comment:
Don't we already have these functions somewhere else? Feels like we're duplicating test code we already have elsewhere, but I don't know where.

Reply:
Yes, but they are not flexible and generic to fit to all use cases. Thats why we have introduced these 2 main code testutil functions - `constructPsbt` and `constructTxnBuilder` and other utility functions. In follow up PRs, we should gradually replace other functions duplicate functions with these 2.